File: 2023/solution/11/solve.py
import functools, re

# Summon magic
__import__('sys').path.append(str(__import__('pathlib').Path(__file__).parents[2]))
from magic import *


# Reading in data
@memoize
def read():
    factor = None
    galaxies = []
    with open_input() as file:
        for i, line in enumerate(file):
            for match in re.finditer(r'#', line):
                j = match.start()
                galaxies.append((i, j))
            if match := re.match(r'\d+', line):
                factor = int(match[0])
    return galaxies, factor


# Solution to part one
@analyze
def solve_one(galaxies, factor):
    @functools.cache
    def get_empty(dim):
        find_extremum = lambda ext, dim: ext(galaxy[dim] for galaxy in galaxies)
        row_min = find_extremum(min, dim)
        row_max = find_extremum(max, dim)
        empty = []
        for row in range(row_min + 1, row_max):
            for galaxy in galaxies:
                if galaxy[dim] == row:
                    break
            else:
                empty.append(row)
        return empty

    def get_dist(galaxy0, galaxy1):
        i0, i1 = extract_ordered(galaxy0, galaxy1, 0)
        j0, j1 = extract_ordered(galaxy0, galaxy1, 1)
        dist = (i1 - i0) + (j1 - j0)
        rows_empty = get_empty(0)
        cols_empty = get_empty(1)
        dist += sum(i0 < row < i1 for row in rows_empty)
        dist += sum(j0 < col < j1 for col in cols_empty)
        return dist

    def extract_ordered(galaxy0, galaxy1, dim):
        i0, i1 = galaxy0[dim], galaxy1[dim]
        if i0 > i1:
            i0, i1 = i1, i0
        return i0, i1

    return sum(
        get_dist(galaxy0, galaxy1)
        for index, galaxy0 in enumerate(galaxies)
        for galaxy1 in galaxies[index + 1 :]
    )


# Solution to part two
@analyze
def solve_two(galaxies, factor):
    if factor is None:
        factor = 1_000_000

    @functools.cache
    def get_empty(dim):
        find_extremum = lambda ext, dim: ext(galaxy[dim] for galaxy in galaxies)
        row_min = find_extremum(min, dim)
        row_max = find_extremum(max, dim)
        empty = []
        for row in range(row_min + 1, row_max):
            for galaxy in galaxies:
                if galaxy[dim] == row:
                    break
            else:
                empty.append(row)
        return empty

    def get_dist(galaxy0, galaxy1):
        i0, i1 = extract_ordered(galaxy0, galaxy1, 0)
        j0, j1 = extract_ordered(galaxy0, galaxy1, 1)
        dist = (i1 - i0) + (j1 - j0)
        rows_empty = get_empty(0)
        cols_empty = get_empty(1)
        expansion = 0
        expansion += sum(i0 < row < i1 for row in rows_empty)
        expansion += sum(j0 < col < j1 for col in cols_empty)
        return dist + (factor - 1) * expansion

    def extract_ordered(galaxy0, galaxy1, dim):
        i0, i1 = galaxy0[dim], galaxy1[dim]
        if i0 > i1:
            i0, i1 = i1, i0
        return i0, i1

    return sum(
        get_dist(galaxy0, galaxy1)
        for index, galaxy0 in enumerate(galaxies)
        for galaxy1 in galaxies[index + 1 :]
    )


# Solve
print(solve_one(*read()))
print(solve_two(*read()))

File: 2024/solution/09/solve.cpp
#include <algorithm>
#include <list>

#include "../../magic.hpp"

using Int = std::int64_t;
struct Block {
    static constexpr Int id_space = -1;
    Int size;
    Int id = id_space;
    bool is_file() const { return id != id_space; }
    void free() { id = id_space; }
};

// Reading in data
std::list<Block> read() {
    std::list<Block> diskmap;
    for (const auto& line : LineReader("input")) {
        for (auto [twice_id, c] : std::views::enumerate(line)) {
            Int size = c - '0';
            if (size == 0) {
                continue;
            }
            if (twice_id & 1) {
                diskmap.push_back({size});
            } else {
                diskmap.push_back({size, static_cast<Int>(twice_id / 2)});
            }
        }
    }
    return diskmap;
}

// Solution to part one
Int solve_one(std::list<Block>& diskmap) {
    for (auto it_space = diskmap.begin(); it_space != diskmap.end(); it_space++) {
        Block space = *it_space;
        if (space.is_file()) {
            continue;
        }
        for (auto it_file = --diskmap.end(); it_file != it_space; it_file--) {
            Block file = *it_file;
            if (not file.is_file()) {
                continue;
            }
            if (it_space == --it_file) {
                // Final move
                it_space = diskmap.erase(it_space, diskmap.end());
                it_space = diskmap.insert(it_space, file);
                break;
            }
            ++it_file;
            // Pop space and file block
            it_space = diskmap.erase(it_space);
            it_file = diskmap.erase(it_file);
            // Insert file blocks
            Int size = std::min(space.size, file.size);
            it_space = diskmap.insert(it_space, {size, file.id});
            if (size < file.size) {
                it_file = diskmap.insert(it_file, {file.size - size, file.id});
            }
            // Insert space block
            if (size < space.size) {
                it_space = --diskmap.insert(++it_space, {space.size - size});
            }
            break;
        }
    }
    Int checksum = 0;
    Int index = 0;
    for (const auto& block : diskmap) {
        checksum += block.id * block.size * (2 * index + block.size - 1) / 2;
        index += block.size;
    }
    return checksum;
}

// Solution to part two
Int solve_two(std::list<Block>& diskmap) {
    for (auto it_file = --diskmap.end(); it_file != diskmap.begin(); it_file--) {
        Block file = *it_file;
        if (not file.is_file()) {
            continue;
        }
        for (auto it_space = diskmap.begin(); it_space != it_file; it_space++) {
            Block space = *it_space;
            if (space.is_file()) {
                continue;
            }
            if (space.size < file.size) {
                continue;
            }
            it_space = diskmap.erase(it_space);
            it_space = diskmap.insert(it_space, file);
            Int size = space.size - file.size;
            if (size > 0) {
                it_space = --diskmap.insert(++it_space, {size});
            }
            it_file->free();
            break;
        }
    }
    Int checksum = 0;
    Int index = 0;
    for (const auto& block : diskmap) {
        if (block.is_file()) {
            checksum += block.id * block.size * (2 * index + block.size - 1) / 2;
        }
        index += block.size;
    }
    return checksum;
}

// Solve
int main() {
    analyze(read, solve_one);
    analyze(read, solve_two);
    return 0;
}
